ecowitt2mqtt/runtime.py

```diff
@@ -67,6 +67,15 @@ def __init__(self, ecowitt: Ecowitt) -> None:
             )
         )
 
+        @app.get("/health/liveness")
+        async def _async_liveness():
+            return { "status": "Online"}
+
+        # TODO: check MQTT is reachable or return 503
+        @app.get("/health/readiness")
+        async def _async_readiness():
+            return { "status": "Ready"}
```

Comment:
I'd rather not merge `TODO` code. Can we wire the real logic in?

Reply:
Yes i can do i just wanted to get this up here for comments earlier rather than later.

This check should effectively ping the mqtt server and if it is unreachable return a 503 which would put the pod in crash loop until the mqtt service is reachable. 

In saying that it is also correct for you app to 'do nothing' and not crash as it does currently, i prefer the crash loop as it makes it obvious that something is wrong.

Reply:
interestingly this code is never reached if i cant connect to the mqtt server.

Reply:
Hmm. The MQTT "loop" (client connection + publishing payloads) happens in a unique task per MQTT broker and they are all separated from the REST API server task, which is where this endpoint is defined. Can you share more details of what you're noticing?

Reply:
basically if I define the MQTT server but not user name and password then my container will never start as the liveness / readiness probe never gets a response. 
It seems your initial connection on startup to MQTT must be a blocking one and not async

`test-ecowitt2mqtt-6fb5c695cf-cx5vm   0/1     CrashLoopBackOff   47         158m`
```
2022-09-02 21:22:38,410 | ecowitt2mqtt | INFO | Starting ecowitt2mqtt (version 2022.08.3)
2022-09-02 21:22:38,434 | ecowitt2mqtt | ERROR | There was an MQTT error: [code:5] The connection was refused.
2022-09-02 21:22:38,436 | ecowitt2mqtt | INFO | Attempting MQTT reconnection in 1 seconds (attempt 1)
2022-09-02 21:22:39,446 | ecowitt2mqtt | ERROR | There was an MQTT error: [code:5] The connection was refused.
2022-09-02 21:22:39,446 | ecowitt2mqtt | INFO | Attempting MQTT reconnection in 4 seconds (attempt 2)
```

Reply:
A simple test should be:
* use this branch to make a docker container, 
* start the container with a vaild mqtt server which requires a password but do not provide password or username. `docker run --rm ..... --name ecotest`
* in a seperate terminal `docker logs ecotest`
* in another terminal `curl -i localhost:8080/health/liveness`

Do the same as above with valid creds and all will work as expected.

Reply:
Something's going on on your end. I just spun up a fresh MQTT broker that doesn't require auth, and `ecowitt2mqtt` connected just fine.

> It seems your initial connection on startup to MQTT must be a blocking one and not async

That isn't correct: as I mentioned before, the MQTT loop starts up in a separate task (an [`asyncio.Task`](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task), in particular) and doesn't block anything. Your logs make me guess that your MQTT broker isn't configured to accept anonymous connections (i.e., those without authentication).

Reply:
`a fresh MQTT broker that doesn't require auth`
No you need to spin up a broker that REQUIRES auth but do not provide the auth details to ecowitt2mqtt

Reply:
Ah, okay – I now see what you're saying.

I just pulled your PR and spun up a fresh instance. I didn't spin up an MQTT broker (which, for the purposes of this test, is effectively the same as not being able to connect to one). I attempted to publish a payload and, as expected, I got log messages about attempting to retry the broker connection:

```
✦ ❯ ecowitt2mqtt --mqtt-broker=127.0.0.1 --mqtt-username=ecowitt --mqtt-password=password --hass-discovery --verbose
2022-09-02 16:32:12,508 | DEBUG | Input CLI options/environment variables: {'mqtt_broker': '127.0.0.1', 'mqtt_username': 'ecowitt', 'mqtt_password': 'password', 'hass_discovery': True, 'verbose': True}
2022-09-02 16:32:12,508 | DEBUG | Configs loaded: <Configs _configs={'default': {'mqtt_broker': '127.0.0.1', 'mqtt_username': 'ecowitt', 'mqtt_password': 'password', 'hass_discovery': True, 'verbose': True, 'hass_discovery_prefix': 'homeassistant', 'mqtt_port': 1883, 'output_unit_system': 'imperial', 'endpoint': '/data/report', 'default_battery_strategy': <BatteryStrategy.BOOLEAN: 'boolean'>, 'mqtt_tls': False, 'input_unit_system': 'imperial', 'port': 8080, 'diagnostics': False, 'raw_data': False, 'battery_override': {}, 'disable_calculated_data': False, 'mqtt_retain': False}}
2022-09-02 16:32:12,511 | INFO | Starting ecowitt2mqtt (version 2022.08.5)
2022-09-02 16:32:12,512 | DEBUG | Starting runtime
2022-09-02 16:32:20,664 | DEBUG | Received data from an Ecowitt device: {'PASSKEY': 'xxxxxxxx', 'stationtype': 'EasyWeatherPro_V5.0.8', 'runtime': '0', 'dateutc': '2022-08-28 03:01:32', 'tempinf': '90.0', 'humidityin': '67', 'baromrelin': '29.769', 'baromabsin': '29.769', 'tempf': '90.0', 'humidity': '73', 'winddir': '287', 'windspeedmph': '0.00', 'windgustmph': '0.00', 'maxdailygust': '2.24', 'solarradiation': '88.67', 'uv': '0', 'rainratein': '0.000', 'eventrainin': '0.000', 'hourlyrainin': '0.000', 'dailyrainin': '0.000', 'weeklyrainin': '0.000', 'monthlyrainin': '2.012', 'yearlyrainin': '2.343', 'soilmoisture1': '45', 'pm25_ch1': '102.0', 'pm25_avg_24h_ch1': '117.7', 'wh65batt': '0', 'wh25batt': '0', 'soilbatt1': '1.5', 'pm25batt1': '4', 'freq': '433M', 'model': 'HP2550_Pro_V1.8.0'}
2022-09-02 16:32:20,664 | DEBUG | Creating MQTT loop: ecowitt@127.0.0.1:1883
INFO:     127.0.0.1:58929 - "POST /data/report HTTP/1.1" 204 No Content
2022-09-02 16:32:20,666 | ERROR | There was an MQTT error: [Errno 61] Connection refused
2022-09-02 16:32:20,667 | INFO | Attempting MQTT reconnection in 1 seconds (attempt 1)
2022-09-02 16:32:21,668 | ERROR | There was an MQTT error: [Errno 61] Connection refused
2022-09-02 16:32:21,669 | INFO | Attempting MQTT reconnection in 4 seconds (attempt 2)
2022-09-02 16:32:25,671 | ERROR | There was an MQTT error: [Errno 61] Connection refused
2022-09-02 16:32:25,671 | INFO | Attempting MQTT reconnection in 9 seconds (attempt 3)
```

In the middle of this, I fired a `GET` to `/health/liveness` and the request worked fine:

```
2022-09-02 16:32:34,672 | ERROR | There was an MQTT error: [Errno 61] Connection refused
2022-09-02 16:32:34,672 | INFO | Attempting MQTT reconnection in 16 seconds (attempt 4)
INFO:     127.0.0.1:58937 - "GET /health/liveness HTTP/1.1" 200 OK
2022-09-02 16:32:50,674 | ERROR | There was an MQTT error: [Errno 61] Connection refused
2022-09-02 16:32:50,674 | INFO | Attempting MQTT reconnection in 25 seconds (attempt 5)
```

Request:

```
GET /health/liveness HTTP/1.1
Content-Type: application/x-www-form-urlencoded; charset=utf-8
Host: 127.0.0.1:8080
Connection: close
User-Agent: Paw/3.4.0 (Macintosh; OS X/12.5.1) GCDHTTPRequest
```

Response:

```
HTTP/1.1 200 OK
date: Fri, 02 Sep 2022 22:32:48 GMT
server: uvicorn
content-length: 19
content-type: application/json
Connection: close

{"status":"Online"}
```

So, you should still be able to query other endpoints even if the MQTT broker connection isn't established.

Reply:
strange.. i wonder why kubes isnt doing that.. maybe i'm being silly and using an old image.

Reply:
do you restart the app if you cant connect?

Reply:
No – as far as this task is concerned, we merely continue trying to connect (with an exponentially-increasing backoff timeout):

https://github.com/bachya/ecowitt2mqtt/blob/8992fa8c58f040bacc26d36bf55185785ea35c8d/ecowitt2mqtt/runtime.py#L118-L128

Reply:
yes I am a muppet.. was using your image with my code (i'd reverted the values file to commit and push DoH!!!)

Reply:
to save myself some time working it out, i'll just ask, (I'm not hugely familiar with asyncio) how can I get a state back to determine if you successfully connected to mqtt. Is there a var i can check is not None or something?

Reply:
There isn't currently a way for code outside the task to know whether the MQTT connection is active. I think storing a private boolean to track this is fine.

Reply:
Actually, I don't think tracking a single boolean is going to be sufficient. https://github.com/bachya/ecowitt2mqtt/pull/264 introduced the possibility of more than one MQTT connection if the user is sending data to multiple brokers – so, probably need some more thinking on what "are we connected?" means.

Reply:
have a look and see what i've got now if you could please. see if that works

Reply:
Problem i see with multiple mqtt is what is the correct response.. Is it acceptable to be classed as 'working' if a single mqtt is up or should all be up or some combo.

Reply:
My use case is a single MQTT so assuming what i have here works could we pass the PR on that basis and if someone raises an issue then fix the use case of multiple MQTT?

I am thinking multiple would be better in a different PR anyway as quickly thinking about it:
1. We need to handle identifying which instance is down
2. We need to provide some way for the user to say 'minimum number of MQTT should be up'
3. List of critical MQTT servers (in case specific ones are allowed to fail and specific ones should be online' (could potentially replace item 2 above if users specifies a list)

Reply:
I think that's very reasonable as a first pass (and a good strategy for figuring out whether something more complex is needed in the future). 👍🏻